stl/inc/ranges

```diff
@@ -22,15 +25,461 @@ _STL_DISABLE_CLANG_WARNINGS
+            }
+        }
+
+        // clang-format off
+        _Semiregular_box& operator=(const _Semiregular_box&) noexcept
+            requires copyable<_Ty> && is_trivially_copy_assignable_v<_Ty> = default;
+        // clang-format on
+
+        _Semiregular_box& operator=(const _Semiregular_box& _That) noexcept(is_nothrow_copy_constructible_v<_Ty>&&
+                is_nothrow_copy_assignable_v<_Ty>) /* strengthened */ requires copyable<_Ty> {
+            if (_Engaged) {
+                if (_That._Engaged) {
+                    _Val = _That._Val;
+                } else {
+                    _Val.~_Ty();
+                    _Engaged = false;
+                }
+            } else {
+                if (_That._Engaged) {
+                    _Construct_in_place(_Val, _That._Val);
+                    _Engaged = true;
+                } else {
+                    ; // nothing to do
+                }
+            }
+
+            return *this;
+        }
+
+        _Semiregular_box& operator=(const _Semiregular_box& _That) noexcept(is_nothrow_copy_constructible_v<_Ty>) {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            if (_That._Engaged) {
+                _Construct_in_place(_Val, _That._Val);
+                _Engaged = true;
+            }
+
+            return *this;
+        }
+
+        // clang-format off
+        _Semiregular_box& operator=(_Semiregular_box&&) noexcept
+            requires movable<_Ty> && is_trivially_move_assignable_v<_Ty> = default;
+        // clang-format on
+
+        _Semiregular_box& operator=(_Semiregular_box&& _That) noexcept(is_nothrow_move_constructible_v<_Ty>&&
+                is_nothrow_move_assignable_v<_Ty>) /* strengthened */ requires movable<_Ty> {
+            if (_Engaged) {
+                if (_That._Engaged) {
+                    _Val = _STD move(_That._Val);
+                } else {
+                    _Val.~_Ty();
+                    _Engaged = false;
+                }
+            } else {
+                if (_That._Engaged) {
+                    _Construct_in_place(_Val, _STD move(_That._Val));
+                    _Engaged = true;
+                } else {
+                    ; // nothing to do
+                }
+            }
+
+            return *this;
+        }
+
+        _Semiregular_box& operator=(_Semiregular_box&& _That) noexcept(is_nothrow_move_constructible_v<_Ty>) {
+            if (_Engaged) {
+                _Val.~_Ty();
+                _Engaged = false;
+            }
+
+            if (_That._Engaged) {
+                _Construct_in_place(_Val, _STD move(_That._Val));
+                _Engaged = true;
+            }
+
+            return *this;
+        }
+
+        constexpr explicit operator bool() const noexcept {
+            return _Engaged;
+        }
+
+        constexpr _Ty& operator*() noexcept {
+            _STL_INTERNAL_CHECK(*this);

```

Comment:
Would it be clearer if this would simply check `_Engaged`

Reply:
Yes, probably. Checking the user-available syntax makes sense in a user-mode precondition check, but checking the implementation detail directly is better in an internal precondition check like this.